lib/rules/new-ember-data-packages.js

```diff
@@ -0,0 +1,168 @@
+
+/**
+ * This function returns an object like this:
+ * {
+ *   'ember-data/model': {
+ *     default: ['@ember-data/model'],
+ *   },
+ * }
+ */
+function oldDataImportsReducer(acc, mapping) {
+  let obj = Object.create(null);
+  obj[mapping.export] = [mapping.replacement.module, mapping.replacement.export];
+  acc[mapping.module] = obj; // eslint-disable-line no-param-reassign
+
+  return acc;
+}
+
+const OLD_DATA_IMPORTS = MAPPINGS.reduce(oldDataImportsReducer, Object.create(null));
+
+/**
+ * This function returns an object like this:
+ * {
+ *   "DS.Model": {
+ *     global: "DS.Model",
+ *     export: "default",
+ *     localName: "Model",
+ *     module: "@ember-data/model" <- replacement
+ *   }
+ * }
+ */
+function globalsReducer(acc, mapping) {
+  const { global } = mapping;
+  if (global in acc) {
+    return acc;
+  }
+
+  let obj = {
+    global,
+    export: mapping.replacement.export,
+    localName: mapping.localName,
+    module: mapping.replacement.module,
+  };
+
+  acc[global] = obj;
+
+  return acc;
+}
+
+const GLOBALS = MAPPINGS.reduce(globalsReducer, Object.create(null));
+
+//------------------------------------------------------------------------------
+// Rule Definition - Use "Ember Data Packages" from Ember RFC #395
+//------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Use "Ember Data Packages" from Ember RFC #395',
+      category: 'Best Practices',
+      recommended: false,
+      url:
+        'https://github.com/ember-cli/eslint-plugin-ember/tree/master/docs/rules/new-ember-data-packages.md',
+    },
+    fixable: 'code',
+  },
+
+  create(context) {
+    //----------------------------------------------------------------------
+    // Variables
+    //----------------------------------------------------------------------
+
+    const message =
+      'Imports from @ember-data packages should be prefered over imports from ember-data';
+
+    //----------------------------------------------------------------------
+    // Public
+    //----------------------------------------------------------------------
+
+    return {
+      /**
+       * turn `import Model from "ember-data/model"`
+       * into `import Model from "@ember-data/model"`
+       * or
+       * `import attr from "ember-data/attr"`
+       * into `import { attr } from "@ember-data/attr"`
+       */
+      ImportDeclaration(node) {
+        if (node.source.value in OLD_DATA_IMPORTS) {
+          const fix = buildFix(node, OLD_DATA_IMPORTS);
+          context.report({ node, message, fix });
+          return;
+        }
+
+        if (node.source.value.match(/^ember-data/)) {
+          context.report(node, message);
+        }
+      },
+
+      /**
+       * warn against `const { Model } = DS`
+       */
+      VariableDeclarator(node) {
+        // Filter out non-DS variable declarations
+        if (!node.init || node.init.name !== 'DS') return;
```

Comment:
Shouldn't this be tracked from imports as well as checking `DS` as is done now?

Can you add a test case for:

```js
import ED from 'ember-data';

const { Model } = ED;
```

Reply:
Thank you, I've added the test case and saw that your snippet could generate two warnings (instead of one). To do so, I've removed the line you pointed me to. Does it look to you? I'm not 100% sure I have understood what you meant. (I keep this modification in an isolated commit to rework it if needed.)

Reply:
New test looks great! Here is another to add:

```js
import LOL from 'who-knows-but-definitely-not-ember-data';

const { Model } = LOL;
```

I would not expect ^ to error/warn, but I _think_ the current code (as written) would.

Reply:
I've addressed the issue in the last commit: https://github.com/ember-cli/eslint-plugin-ember/pull/450/commits/ca976143e2bd7c1c8d9408b6a423dbc8446922b0